Added Chat Stacking

MystWraith opened the pull request:
# Description
This is pretty simple, it stacks messages to make the chat less cluttered.
## Inspiration / Reference
~~This was heavily derived from Compact Chat ([GitHub](https://github.com/caoimhebyrne/compact-chat), [Modrinth](https://modrinth.com/mod/compact-chat))~~
The above mentioned mod is very similar to many more mods actually. Recode has this feature.
So does [ChatUtils](https://modrinth.com/mod/chatutils)
And [Chat Patches](https://modrinth.com/mod/chatpatches)
And [Advanced Chat](https://modrinth.com/mod/advancedchat)

I did modify it and try to make it more in the format of how this mod is already configured.

# Plans:
- [x] Chat stacking
- [x] Toggle in config

# Icing on the cake:
- [x] Page in config
- [x] Custom colors
- [x] Custom bracket options
  - [x] Enum
  - [x] Text input for custom?

MystWraith committed:
Added chat stacking

This was heavily derived from Compact Chat ([GitHub](https://github.com/caoimhebyrne/compact-chat), [Modrinth](https://modrinth.com/mod/compact-chat))

MystWraith committed:
Options :)

Made the name more consistent and made the check to see if it is enabled.

showierdata9978 commented:
This will take a while to review (when you mark it as ready) because il have to check cc's code too see if any code was copied to comply with the MIT license

MystWraith committed:
Fixed Regex bug

It used to not work with multiple matches (matches beyond x2).

I found out I needed to include the `§` character.

MystWraith commented:
Ah, yeah. I think I copied some bigger things, I'll try to get a list of all things I copied, but I also removed a ton of unnecessary code.

MystWraith commented:
# "Copied" Code
<details><summary>

## ~~`hook`~~ Removed (Overhead)</summary>

```java
public Text compactChatMessage(Text message) {
    // We only use the message without timestamps when comparing to the previous/other messages.
    // This is because the timestamps are not part of the message, and thus should not be used for comparison.
    // However, we do want to keep the timestamps in the message, so we can't remove them from the message.
    // See GitHub issue #23 for more information.
    var withoutTimestamps = TextUtil.removeTimestamps(message);

    var chatMessage = this.chatMessages.get(withoutTimestamps);
    var previousMessage = this.previousMessage;
    this.previousMessage = withoutTimestamps;

    var shouldIgnoreNonConsecutiveMessage = Configuration.getInstance().onlyCompactConsecutiveMessages && !withoutTimestamps.equals(previousMessage);
    if (chatMessage == null || shouldIgnoreNonConsecutiveMessage || shouldIgnoreCommonSeparator(message)) {
        this.chatMessages.put(withoutTimestamps, new ChatMessage());
        return message;
    }

    // This chat message has occurred before, let's remove the old occurrence(s).
    this.removeMessage(message, chatMessage);
    chatMessage.addOccurrence();

    return chatMessage.modifiedText(message);
}

. . .

public void removeMessage(Text originalMessage, ChatMessage message) {
    var iterator = this.chatHud.compactchat$getMessages().listIterator();
    while (iterator.hasNext()) {
        var chatHudLine = iterator.next();

        // We remove occurrences because we want to remove existing compacted messages too.
        var contentWithoutOccurrences = message.removeTextModifications(chatHudLine.content());
        var textWithoutOccurrences = message.removeTextModifications(originalMessage);

        if (contentWithoutOccurrences.equals(textWithoutOccurrences)) {
            iterator.remove();
            this.chatHud.compactchat$refreshMessages();

            return;
        }
    }
}
```
</details><details><summary>

## ~~`hookExtension` (Interface)~~ Removed (Overhead)</summary>

```java
public interface IChatHudExt {
    List<ChatHudLine> compactchat$getMessages();
    void compactchat$refreshMessages();
    void compactchat$clear();
}
```
</details><details><summary>

## `TextUtil` Collapsed and moved to `Regexs`</summary>

```java
public class TextUtil {
    /**
     * A regex pattern to match a timestamp of HH:mm(:ss?).
     * It can be surrounded by any character.
     */
    private static final String TIMESTAMP_PATTERN = ".?\\d{1,2}:\\d{2}(:\\d{2})*.?";

    public static Text removeSiblings(Text parent, Predicate<Text> predicate) {
        var copy = parent.copy();
        copy.getSiblings().removeIf(predicate);

        return copy;
    }

    /**
     * Most mods add create a new text literal and append the original text to it.
     * This method removes the timestamp from the literal text content.
     */
    public static Text removeTimestamps(Text text) {
        var content = text.getContent();
        if (!(content instanceof LiteralTextContent literalTextContent)) {
            return text;
        }

        var string = literalTextContent.string();
        var withoutTimestamps = string.replaceAll(TIMESTAMP_PATTERN, "");
        if (withoutTimestamps.equals(string)) {
            return text;
        }

        var newText = Text.literal(withoutTimestamps.trim());
        newText.setStyle(newText.getStyle());
        newText.getSiblings().addAll(text.getSiblings());

        return newText;
    }
}
```
</details><details><summary>

## `Mixin` Interface removed and compacted</summary>

```java
public abstract class ChatHudMixin implements IChatHudExt {
    @Unique
    private final ChatHudHook compactchat$hook = new ChatHudHook(this);

    @Shadow
    @Final
    private List<ChatHudLine> messages;

    @Shadow
    public abstract void reset();

    @Shadow
    public abstract void clear(boolean clearHistory);

    @ModifyVariable(
        method = "addMessage(Lnet/minecraft/text/Text;Lnet/minecraft/network/message/MessageSignatureData;ILnet/minecraft/client/gui/hud/MessageIndicator;Z)V",
        at = @At("HEAD"),
        argsOnly = true
    )
    private Text compactchat$compactChatMessage(Text message, Text parameterMessage, MessageSignatureData data, int ticks, MessageIndicator indicator, boolean refreshing) {
        // If we're refreshing, we have probably already modified the message, therefore we don't want to do anything.
        if (refreshing) {
            return message;
        }

        return this.compactchat$hook.compactChatMessage(message);
    }

    @Inject(method = "clear", at = @At("RETURN"))
    private void compactchat$onClear(boolean clearHistory, CallbackInfo ci) {
        this.compactchat$hook.reset();
    }

    @Override
    public List<ChatHudLine> compactchat$getMessages() {
        return this.messages;
    }

    @Override
    public void compactchat$refreshMessages() {
        this.reset();
    }

    @Override
    public void compactchat$clear() {
        this.clear(false);
    }
}
```
</details>

## Note
Some of these are really small and some I came up with identical solutions.
I tried to use this as a reference and not copy verbatim and without thought. I tried my best to understand this code and had to do some debugging anyway.
I did move some of the code into different places to make it match this mod's format.

## Legal Aknowledgement
I understand that stealing and plagiarizing is wrong. I know I used the mod mentioned in the original post as heavy inspiration and reference. I understand that copyright and legal ownership belongs to the original owners. I understand that this decision to mirror someone else's code may lead to the cancelation of this pull request.
Apologies in advance for any headache I may have caused and any legal issues I may have stirred.

showierdata9978 commented:
all that needs to be done *prob* is add comments like this

```java
/*
Parts of this file are Copyright (c) 2023 Caoimhe Byrne

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
```

another thing tho; i dont want to be responsible for maintaining it

showierdata9978 commented:
iirc recode has this built in too

MystWraith commented:
It does, but I wanted to make it more compact, bc I want to make a lightweight modpack/installation using as few unneeded mods and suck as possible.

MystWraith commented:
I'll try to come up with a better way to do this without stolen code. I know it won't reopen it, but I still want to learn and improve my programing skill

MystWraith committed:
Hook stupid

MystWraith commented:
Btw, I should probably make a note, I think I was giving the other mod a bit too much credit. There was a lot of stuff I looked up and figure out on my own (to modify the `ChatHud` class and use an `iterator` to remove the items), the primary reference was how to make the removal actually stick.

The only thing that I don't understand right now is what a `Sibling` is in this context. I think it is changes? It does not seem like a useful name. I'll try to look for more info on them.

MystWraith commented:
I just checked, and yes `Siblings` are just changes made to the `Text`. I don't know why it is called this, but at least I now know what it is.

MystWraith commented:
Yeah, I gave way to much credit. I removed the `hook` bc it was stupid. I also removed a `class` they were using because it was entirely useless. Most of the things I "stole" are common methods to do certain actions like "Remove item from List Java" or "Sort List with Predicate" (which you have seen me do before).

MystWraith committed:
Removed overhead

There was so much garbage from the mod I copied. It's useless code.

MystWraith commented:
Thanks @showierdata9978 for letting me continue this.

MystWraith committed:
Added Bracket changing :smile:

Took some weird stuffs bc it can change and therefore cannot be static

showierdata9978 commented:
You should add a field in the constructor for the actual bracket symbol instead of a switch case

MystWraith commented:
>   * [x]  Text input for custom?

That's in the plans! I'll get working on it. I haven't used ModMenu a whole lot, so I gotta get used to it.

showierdata9978 commented:
You can also just name it `BracketEnum`

MystWraith committed:
Made fun ModMenu Stuff

Added "Custom" bracket area, and made much fun stuff

showierdata9978 commented:
Style: Please use a separate config class for it if you create a new category.

It just makes it less cluttered

MystWraith commented:
What do you mean? Create a new function?

showierdata9978 commented:
New class for the stacking config

Look at
https://github.com/showierdata9978/Pickaxe-mod/blob/main/src/main/java/tech/showierdata/pickaxe/config/ItemConfig.java
for an example

On Thu, Aug 31, 2023, 1:08 PM MystWraith ***@***.***> wrote:

> What do you mean? Create a new function?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/showierdata9978/Pickaxe-mod/pull/32#issuecomment-1701529710>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQHW4P7OKN5K7QX2QAAU26TXYDHJXANCNFSM6AAAAAA4A5TMMY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

MystWraith committed:
Per request

Had to double check that it didn't break

MystWraith commented:
The only thing I couldn't figure out is how to make the previews appear in the text editor for the prefix and suffix, but I put it in the description, so it's not the worst.

MystWraith committed:
Cleaning / Docstrings

MystWraith committed:
Oops

I forgot some things

MystWraith committed:
Much nicer

Changed the config style, I think it looks a lot better.
Should also be easier to use

MystWraith commented:
Bc now it looks good too

showierdata9978 commented:
Il test this as soon as i get home
